src/hooks/useSourceExpansion.js

```diff
@@ -0,0 +1,50 @@
+import {useReducer, useEffect, useMemo} from 'react';
+import {expandFromRawCode} from '../utils';
+
+const useAppendableCollection = () => {
+    const update = (collection, {type, value}) => {
+        switch (type) {
+            case 'clear':
+                return collection.length ? [] : collection;
+            case 'push':
+                return [...collection, value];
+            default:
+                return collection;
+        }
+    };
+
+    const [collection, dispatch] = useReducer(update, []);
+    return {
+        collection,
+        clear() {
+            dispatch({type: 'clear'});
+        },
+        push(value) {
+            dispatch({value, type: 'push'});
+        },
+    };
+};
+
+export default (hunks, oldSource) => {
+    const {collection: expandedRanges, clear, push} = useAppendableCollection();
+    useEffect(clear, [hunks, oldSource]);
+    const linesOfOldSource = useMemo(() => (oldSource || '').split('\n'), [oldSource]);
```

Comment:
I find smaller `useMemo` is a good practice to cache computation in more rational pieces.